libraries/program-error-derive/README.md

```diff
@@ -0,0 +1,166 @@
+    fn fmt(&self, __formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+        #[allow(unused_variables, deprecated, clippy::used_underscore_binding)]
+        match self {
+            ExampleError::MintHasNoMintAuthority {} => {
+                __formatter.write_fmt(format_args!("Mint has no mint authority"))
+            }
+            ExampleError::IncorrectMintAuthority {} => {
+                __formatter
+                    .write_fmt(
+                        format_args!(
+                            "Incorrect mint authority has signed the instruction"
+                        ),
+                    )
+            }
+        }
+    }
+}
+#[allow(non_upper_case_globals, unused_qualifications)]
+const _IMPL_NUM_FromPrimitive_FOR_ExampleError: () = {
+    #[allow(clippy::useless_attribute)]
+    #[allow(rust_2018_idioms)]
+    extern crate num_traits as _num_traits;
+    impl _num_traits::FromPrimitive for ExampleError {
+        #[allow(trivial_numeric_casts)]
+        #[inline]
+        fn from_i64(n: i64) -> Option<Self> {
+            if n == ExampleError::MintHasNoMintAuthority as i64 {
+                Some(ExampleError::MintHasNoMintAuthority)
+            } else if n == ExampleError::IncorrectMintAuthority as i64 {
+                Some(ExampleError::IncorrectMintAuthority)
+            } else {
+                None
+            }
+        }
+        #[inline]
+        fn from_u64(n: u64) -> Option<Self> {
+            Self::from_i64(n as i64)
+        }
+    }
+};
+#[automatically_derived]
+impl ::core::marker::StructuralPartialEq for ExampleError {}
+#[automatically_derived]
+impl ::core::cmp::PartialEq for ExampleError {
+    #[inline]
+    fn eq(&self, other: &ExampleError) -> bool {
+        let __self_tag = ::core::intrinsics::discriminant_value(self);
+        let __arg1_tag = ::core::intrinsics::discriminant_value(other);
+        __self_tag == __arg1_tag
+    }
+}
+impl From<ExampleError> for solana_program::program_error::ProgramError {
+    fn from(e: ExampleError) -> Self {
+        solana_program::program_error::ProgramError::Custom(e as u32)
+    }
+}
+impl<T> solana_program::decode_error::DecodeError<T> for ExampleError {
+    fn type_of() -> &'static str {
+        "ExampleError"
+    }
+}
+impl solana_program::program_error::PrintProgramError for ExampleError {
+    fn print<E>(&self)
+    where
+        E: 'static + std::error::Error + solana_program::decode_error::DecodeError<E>
+            + solana_program::program_error::PrintProgramError
+            + num_traits::FromPrimitive,
+    {
+        match self {
+            ExampleError::MintHasNoMintAuthority => {
+                ::solana_program::log::sol_log("Mint has no mint authority")
+            }
+            ExampleError::IncorrectMintAuthority => {
+                ::solana_program::log::sol_log(
+                    "Incorrect mint authority has signed the instruction",
+                )
+            }
+        }
+    }
+}
```

Comment:
It's these last three that are the most annoying to remember to add. Rather than just one macro that adds everything, how about several macros for each of these?

It looks like `impl From<ExampleError> for solana_program::program_error::ProgramError` requires `num_derive::FromPrimitive` to do that `e as u32` part.

And it looks like `impl solana_program::program_error::PrintProgramError` requires `thiserror::Error` and `DecodeError`.

So maybe there's four separate macros, with an option to do the full thing. ie.

```
#[print_program_error]
#[decode_error]
#[into_program_error]
#[Derive(...)]
pub enum MyError { }
```

but then you have `#[spl_program_error]` which just does all of them. That way you can do whatever you wish.  What do you think?

Side note: we might even consider adding them directly to https://github.com/solana-labs/solana/tree/master/sdk/macro as separate macros.